### PR TITLE
Add ability to push logs to ElasticSearch

### PR DESCRIPTION
**Pull request by bitwiseman**
Log to  ElasticSearch using REST API.  Updated readme, added basic tests, could use more.

### CODE REVIEW

File: src/main/java/jenkins/plugins/logstash/persistence/ElasticSearchDao.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import java.io.PrintStream;
+
+/**
+ * Elastic Search Data Access Object.
+ *
+ * @author Liam Newman
+ * @since 1.0.5
```
Comment (R-Gerard) [inline src/main/java/jenkins/plugins/logstash/persistence/ElasticSearchDao.java]: The next version of the plugin will be 1.0.4.

File: pom.xml [hunk 2]
```diff
@@ -43,6 +43,12 @@
   <url>http://wiki.jenkins-ci.org/display/JENKINS/Logstash+Plugin</url>
 
   <dependencies>
+      <dependency>
+          <groupId>org.apache.httpcomponents</groupId>
+          <artifactId>httpclient</artifactId>
+          <version>4.4</version>
+      </dependency>
```
Comment (R-Gerard) [inline pom.xml]: Minor complaint, but please use two spaces instead of 4 to indent.

File: src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    when(mockHttpClient.execute(mockPost)).thenReturn(mockResponse);
+  }
+
+  @After
+  public void after() throws Exception {
+    verifyNoMoreInteractions(mockClientBuilder);
+    verifyNoMoreInteractions(mockHttpClient);
+    verifyNoMoreInteractions(mockPost);
+    verifyNoMoreInteractions(mockLogger);
+  }
```
Comment (R-Gerard) [inline src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java]: You have some unused imports that can be cleaned up:

import org.apache.http.entity.StringEntity;
import org.mockito.Matchers;
import java.net.SocketException;

File: src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    // Unit under test
+    dao = new ElasticSearchDao("localhost", 8200, "/jenkins//logstash/", "username", "password");
+
+    // Verify results
+    assertEquals("Wrong host name", "localhost", dao.host);
+    assertEquals("Wrong port", 8200, dao.port);
+    assertEquals("Wrong key", "/jenkins//logstash/", dao.key);
+    assertEquals("Wrong password", "password", dao.password);
+  }
+}
```
Comment (R-Gerard) [inline src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java]: Can you add some tests for the `push` method in ElasticSearchDao? (A happy path test at the very minimum.) This will help catch any errors that might be introduced when others modify your code.

File: src/main/java/jenkins/plugins/logstash/persistence/ElasticSearchDao.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+  @Override
+  public long push(String data, PrintStream logger) {
+    CloseableHttpClient httpClient = null;
+    CloseableHttpResponse response = null;
+    try {
+      httpClient = clientBuilder.build();
+      HttpPost postRequest = new HttpPost(uri);
+      StringEntity input = new StringEntity(data, "UTF-8");
+      input.setContentType("application/json");
```
Comment (R-Gerard) [inline src/main/java/jenkins/plugins/logstash/persistence/ElasticSearchDao.java]: Minor nitpick: there's constants defined for these strings in java.nio.charset.StandardCharsets and org.apache.http.entity.ContentType:

`StandardCharsets.UTF_8`
`ContentType.APPLICATION_JSON`

**Comment by R-Gerard** [comment]
Thanks for your contribution. The pull request looks good. I've done a small amount of manual testing with my local Jenkins and it seems fine. Please amend your commit with some more unit tests and I'll merge the pull request.

File: src/main/java/jenkins/plugins/logstash/persistence/ElasticSearchDao.java (hunk 5 above)
Comment (bitwiseman) [inline src/main/java/jenkins/plugins/logstash/persistence/ElasticSearchDao.java]: Sure thing.

File: src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java (hunk 4 above)
Comment (bitwiseman) [inline src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java]: I wasn't sure how to do this with the mocks.  I'll take a swing at it but you'll need to take a look when I'm done.

File: src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java (hunk 3 above)
Comment (bitwiseman) [inline src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java]: Of course.

File: pom.xml (hunk 2 above)
Comment (bitwiseman) [inline pom.xml]: No problem.

**Comment by bitwiseman** [comment]
Is there anything in the Jenkins pull request document that you wanted me to apply to this change? It all seems pretty reasonable and I'm not seeing any point where I'm not doing what it suggests.

File: src/main/java/jenkins/plugins/logstash/persistence/ElasticSearchDao.java (hunk 1 above)
Comment (bitwiseman) [inline src/main/java/jenkins/plugins/logstash/persistence/ElasticSearchDao.java]: Right, sorry.

File: src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java (hunk 4 above)
Comment (R-Gerard) [inline src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java]: No problem.

**Comment by R-Gerard** [comment]
Nope, this seems fine. As a rule of thumb it's always a good idea to do some manual testing on a real instance of Jenkins (i.e. not using `mvn hpi:run`), look at the Jenkins service logs, etc.

**Comment by bitwiseman** [comment]
Should I squash these into one commit?

File: src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    assertEquals("Wrong auth", 1, post.getHeaders("Authorization").length);
+    assertEquals("Wrong auth value", "Basic dXNlcm5hbWU6cGFzc3dvcmQ=", post.getHeaders("Authorization")[0].getValue());
+
+
+    assertEquals("Wrong content type", entity.getContentType().getValue(), ContentType.APPLICATION_JSON.toString());
+    assertTrue("Wrong content class", entity instanceof StringEntity);
+
+    ByteArrayOutputStream stream = new ByteArrayOutputStream();
+    entity.writeTo(stream);
+    assertEquals("Wrong content", stream.toString(CharEncoding.UTF_8), "{ 'foo': 'bar' }");
```
Comment (R-Gerard) [inline src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java]: `assertEquals("Wrong content", stream.toString(CharEncoding.UTF_8), json);`

File: src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    HttpEntity entity = post.getEntity();
+
+    assertEquals("Wrong uri", new URI("http://localhost:8200/jenkins/logstash") , post.getURI());
+    assertEquals("Wrong auth", 0, post.getHeaders("Authorization").length);
+    assertEquals("Wrong content type", entity.getContentType().getValue(), ContentType.APPLICATION_JSON.toString());
+    assertTrue("Wrong content class", entity instanceof StringEntity);
+
+    ByteArrayOutputStream stream = new ByteArrayOutputStream();
+    entity.writeTo(stream);
+    assertEquals("Wrong content", stream.toString(CharEncoding.UTF_8), "{ 'foo': 'bar' }");
```
Comment (R-Gerard) [inline src/test/java/jenkins/plugins/logstash/persistence/ElasticSearchDaoTest.java]: `assertEquals("Wrong content", stream.toString(CharEncoding.UTF_8), json);`

**Comment by R-Gerard** [comment]
Looks good. Go ahead and squash the commits and I'll merge it.
